### PR TITLE
Compute facial adjacency with numpy

### PR DESCRIPTION
**Pull request by majosm**
Rewrites `_compute_facial_adjacency_from_vertices` using numpy in order to try to fix illinois-ceesd/mirgecom#202.

### CODE REVIEW

File: meshmode/mesh/__init__.py [hunk 1]
```diff
@@ -1052,6 +1052,43 @@ def _boundary_tag_bit(boundary_tags, btag_to_index, boundary_tag):
...
 
+class _FlatFacialAdjacencyData:
+    """
+    Data structure for intermediate storage of facial adjacency data. Each attribute
+    is a :class:`numpy.ndarray` containing data for each stored face and its
+    adjacent neighbor.
+
+    .. attribute:: elements
+
+        The group-relative element index. (Defau
```
Comment (matthiasdiener) [inline meshmode/mesh/__init__.py:1063]: missing text here.

Comment (majosm) [reply meshmode/mesh/__init__.py:1063]: Whoops. Got distracted. 🙂

File: meshmode/mesh/__init__.py [hunk 2]
```diff
@@ -1063,140 +1100,126 @@ def _compute_facial_adjacency_from_vertices(groups, boundary_tags,
...
     for igrp, grp in enumerate(groups):
-        for fid, face_vertex_indices in enumerate(grp.face_vertex_indices()):
-            all_fvi = grp.vertex_indices[:, face_vertex_indices]
+        for fid, ref_fvi in enumerate(grp.face_vertex_indices()):
+            nvertices = len(ref_fvi)
+            face_nr_bases[igrp, fid] = nfaces_of_size.setdefault(nvertices, 0)
+            nfaces_of_size[nvertices] += grp.nelements
+
+    face_vertex_indices = {}
+    face_ids = {}
```
Comment (inducer) [inline meshmode/mesh/__init__.py:1115]: Add a comment documenting the structure of these.

File: meshmode/mesh/__init__.py [hunk 3]
```diff
@@ -1063,140 +1100,126 @@ def _compute_facial_adjacency_from_vertices(groups, boundary_tags,
...
+        vertex_indices_increasing = np.sort(vertex_indices, axis=0)
+        order = np.lexsort(vertex_indices_increasing)
+        diffs = np.diff(vertex_indices_increasing[:, order], axis=1)
+        match_indices, = (~np.any(diffs, axis=0)).nonzero()
+        matching_faces = (order[match_indices], order[match_indices+1])
+        adj_indices = np.empty(nfaces, dtype=element_id_dtype)
+        adj_indices[:] = -1
+        adj_indices[matching_faces[0]] = matching_faces[1]
+        adj_indices[matching_faces[1]] = matching_faces[0]
+        adjacent_face_indices[nvertices] = adj_indices
```
Comment (inducer) [inline meshmode/mesh/__init__.py:1148]: Give each of these blocks at least a one-liner comment on what they do/what they build.

File: meshmode/mesh/__init__.py [hunk 4]
```diff
@@ -1063,140 +1100,126 @@ def _compute_facial_adjacency_from_vertices(groups, boundary_tags,
...
+    for igrp, grp in enumerate(groups):
+        for fid, ref_fvi in enumerate(grp.face_vertex_indices()):
+            nvertices = len(ref_fvi)
+            face_nr_base = face_nr_bases[igrp, fid]
+            grp_fvi = grp.vertex_indices[:, ref_fvi]
+            vertex_indices = face_vertex_indices[nvertices]
+            ids = face_ids[nvertices]
+            vertex_indices[:, face_nr_base:face_nr_base+grp.nelements] = grp_fvi.T
+            ids[0, face_nr_base:face_nr_base+grp.nelements] = igrp
+            ids[1, face_nr_base:face_nr_base+grp.nelements] = fid
```
Comment (inducer) [inline meshmode/mesh/__init__.py:1130]: I find it hard to read code with aliases that are written to. I'd take a little bit of redundant code over that... it makes it easier to see/Ctrl-F where a given variable is written.

File: meshmode/mesh/__init__.py [hunk 5]
```diff
@@ -1063,140 +1100,126 @@ def _compute_facial_adjacency_from_vertices(groups, boundary_tags,
...
+    face_nr_bases[:] = -1
     for igrp, grp in enumerate(groups):
-        for fid, face_vertex_indices in enumerate(grp.face_vertex_indices()):
-            all_fvi = grp.vertex_indices[:, face_vertex_indices]
+        for fid, ref_fvi in enumerate(grp.face_vertex_indices()):
+            nvertices = len(ref_fvi)
+            face_nr_bases[igrp, fid] = nfaces_of_size.setdefault(nvertices, 0)
+            nfaces_of_size[nvertices] += grp.nelements
+
+    face_vertex_indices = {}
```
Comment (inducer) [inline meshmode/mesh/__init__.py:1114]: I occurred to me that we can probably get away with just one of these, not one per size (as long as the other entries are filled with `-1` or so). I don't know that that's categorically better though. I figured I'd mention it, but I'm not insisting it be done that way.

File: meshmode/mesh/__init__.py [hunk 6]
```diff
@@ -1063,140 +1100,126 @@ def _compute_facial_adjacency_from_vertices(groups, boundary_tags,
...
+                        for tag in tags:
+                            tag_mask |= boundary_tag_bit(tag)
+                        adj.neighbors[iel] = -((-adj.neighbors[iel]) | tag_mask)
+            istart = fid*grp.nelements
+            iend = (fid+1)*grp.nelements
+            grp_adj.elements[istart:iend] = adj.elements
+            grp_adj.element_faces[istart:iend] = adj.element_faces
+            grp_adj.neighbor_groups[istart:iend] = adj.neighbor_groups
+            grp_adj.neighbors[istart:iend] = adj.neighbors
+            grp_adj.neighbor_faces[istart:iend] = adj.neighbor_faces
```
Comment (inducer) [inline meshmode/mesh/__init__.py:1197]: Add comments to document the algorithmic outline of this, too.

**Review by inducer — COMMENTED** [review]
Thanks for working on this! LGTM aside from needing a few comments.

**Comment by inducer** [comment]
This is nice! Thanks for working on it.

**Pull request merged**